prov/efa/src/efa_av.c

```diff
@@ -384,38 +387,53 @@ struct efa_conn *efa_conn_alloc(struct efa_av *av, struct efa_ep_addr *raw_addr,
 	if (!conn->ah)
 		goto err_release;
 
-	key.ahn = conn->ah->ahn;
-	key.qpn = raw_addr->qpn;
-	/* This is correct since the same address should be mapped to the same ah. */
-	HASH_FIND(hh, av->reverse_av, &key, sizeof(key), reverse_av);
-	if (!reverse_av) {
-		reverse_av = malloc(sizeof(*reverse_av));
-		if (!reverse_av) {
-			errno = FI_ENOMEM;
-			goto err_release;
-		}
-
-		memcpy(&reverse_av->key, &key, sizeof(key));
-		reverse_av->conn = conn;
-		HASH_ADD(hh, av->reverse_av, key,
-			 sizeof(reverse_av->key), reverse_av);
-	} else {
-		reverse_av = NULL;
-	}
-
 	if (av->ep_type == FI_EP_RDM) {
 		err = efa_conn_rdm_init(av, conn);
 		if (err)
 			goto err_release;
 	}
 
+	key.ahn = conn->ah->ahn;
+	key.qpn = raw_addr->qpn;
+	reverse_av_entry = NULL;
+	HASH_FIND(hh, av->reverse_av, &key, sizeof(key), reverse_av_entry);
+	if (reverse_av_entry) {
+		/*
+		 * If we found an existing entry in reverse_av, the peer must
+		 * have the same GID and qpn, but different qkey, which means
+		 * the old peer has been destroied. so we have to release the
+		 * corresponding efa_conn object
+		 */
+		prev_conn = reverse_av_entry->conn;
+		assert(prev_conn);
+		assert(memcmp(prev_conn->ep_addr.raw, conn->ep_addr.raw, EFA_GID_LEN)==0);
+		assert(prev_conn->ep_addr.qpn == conn->ep_addr.qpn);
+		assert(prev_conn->ep_addr.qkey != conn->ep_addr.qkey);
+		EFA_WARN(FI_LOG_AV, "QP reuse detected! Previous qkey: %d Current qkey: %d\n",
+			 prev_conn->ep_addr.qkey, conn->ep_addr.qkey);
+		conn->rdm_peer.prev_qkey = prev_conn->ep_addr.qkey;
```

Comment:
What are you thinking of saving this information for?

Reply:
This will be used to ignore packets from previous peer.